xpc_misc.c

```diff
@@ -114,40 +113,457 @@ xpc_array_destroy(struct xpc_object *dict)
+	}
+	*offset += val_packed_size;
+	return xo;
+};
+
+XPC_INLINE int xpc_pack(struct xpc_object* xo, void* buf, size_t* final_size) {
+	uint8_t* out = buf;
+	__block size_t size = 1;
+
+	if (out)
+		out[0] = xo->xo_xpc_type;
+
+	switch (xo->xo_xpc_type) {
+		case _XPC_TYPE_DICTIONARY: {
+			__block size_t offset = size;
+			xpc_dictionary_apply(xo, ^bool(const char* key, xpc_object_t value) {
+				struct xpc_object* val_xo = value;
+				size_t key_size = strlen(key) + 1;
+				size += key_size;
+				if (out) {
+					strncpy(out + offset, key, key_size - 1);
+					offset += key_size;
+					out[offset - 1] = '\0';
+				}
+				xpc_pack_encode_entry(out, &size, &offset, val_xo);
+				return true;
+			});
+		} break;
+		case _XPC_TYPE_ARRAY: {
+			__block size_t offset = size;
+			xpc_array_apply(xo, ^bool(size_t idx, xpc_object_t value) {
+				struct xpc_object* val_xo = value;
+				xpc_pack_encode_entry(out, &size, &offset, val_xo);
+				return true;
+			});
+		} break;
+		case _XPC_TYPE_BOOL: {
+			if (out)
+				out[0] |= (xpc_bool_get_value(xo) ? 1 : 0) << XPC_PACK_TYPE_BITS;
+		} break;
+		case _XPC_TYPE_DATE:
+		case _XPC_TYPE_INT64: {
+			int64_t val = xo->xo_xpc_type == _XPC_TYPE_DATE ? xpc_date_get_value(xo) : xpc_int64_get_value(xo);
+			int64_t val_le = OSSwapHostToLittleInt64(val);
+			size_t val_bits = xpc_pack_encode_int_bits(&val_le, sizeof(int64_t), false);
+			size += XPC_PACK_INT_EXTRA_BYTES(val_bits);
+			if (out)
+				xpc_pack_encode_int(out, &val_le, sizeof(int64_t));
+		} break;
+		case _XPC_TYPE_UINT64: {
+			uint64_t val = xpc_uint64_get_value(xo);
+			uint64_t val_le = OSSwapHostToLittleInt64(val);
+			size_t val_bits = xpc_pack_encode_int_bits(&val_le, sizeof(uint64_t), false);
+			size += XPC_PACK_INT_EXTRA_BYTES(val_bits);
+			if (out)
+				xpc_pack_encode_int(out, &val_le, sizeof(uint64_t));
+		} break;
+		case _XPC_TYPE_DATA: {
+			size_t len = xpc_data_get_length(xo);
+			size += len;
+			if (out)
+				memcpy(out + 1, xpc_data_get_bytes_ptr(xo), len);
+		} break;
+		case _XPC_TYPE_STRING: {
+			size_t len = xpc_string_get_length(xo);
+			size += len + 1;
+			if (out) {
+				strncpy(out + 1, xpc_string_get_string_ptr(xo), len);
+				out[1 + len] = '\0';
+			}
+		} break;
+		// currently unimplemented
+		//case _XPC_TYPE_UUID: {} break;
+		case _XPC_TYPE_FD: {
+			int val = xo->xo_u.port;
+			int val_le = OSSwapHostToLittleInt32(val);
+			size_t val_bits = xpc_pack_encode_int_bits(&val_le, sizeof(int), false);
+			size += XPC_PACK_INT_EXTRA_BYTES(val_bits);
+			if (out)
+				xpc_pack_encode_int(out, &val_le, sizeof(int));
+		} break;
+		case _XPC_TYPE_DOUBLE: {
+			// this is a rather simplistic implementation that assumes that the internal layout
+			// of a double is the same on sender and reeceiver
+			// that's never a good assumption.
+			// (although from what i can tell, XPC is only used for communication between processes
+			// on the same system, so this shouldn't be a problem)
```

Comment:
That's not true anymore, they now use it to communicate to BridgeOS on T2. See https://duo.com/labs/research/apple-t2-xpc

But we don't have to care about that.

Reply:
Oh, actually we do have to care if we want to support Rosetta, i.e. running PowerPC binaries on the same system.

But then again, you don't trying to match the format Apple uses, right? So we'll have to rewrite this anyway.

Reply:
That link is actually super helpful! I would've preferred to implement it the way Apple does, but I couldn't find anywhere that describes the format (but now that I think about it, I could've written two small C programs on a real mac and sending XPC messages between them and dump the message from `mach_msg`).

> But then again, you don't trying to match the format Apple uses, right? So we'll have to rewrite this anyway.

But doesn't Apple say that the format isn't guaranteed to stay consistent across releases? Much like how you're not supposed to make system calls yourself; Apple doesn't guarantee compatibility between releases so you're supposed to use the API they provide in system libraries. Programs aren't supposed have their own libxpc, so the format we use for messages shouldn't matter. Our libxpc is supposed to be the one doing both the packing and the unpacking for all programs, right? Or did Apple use a different approach (i.e. statically linking their libxpc into programs) in the PowerPC days?

Then again, it doesn't matter to me what we do here, I just want XPC messaging to work for trustd <-> curl. The `nvlist`/`nvpair` approach was broken, and I thought this was the easiest way to fix it.

But yeah, with regard to the `double` encoding/decoding, that's definitely going to have to be reimplemented to be portable.

Reply:
> That link is actually super helpful! I would've preferred to implement it the way Apple does, but I couldn't find anywhere that describes the format

Here are some resources:

* https://github.com/saelo/pwn2own2018/tree/master/libspc
* https://www.fortinet.com/blog/threat-research/a-look-into-xpc-internals--reverse-engineering-the-xpc-objects.html
* http://web.archive.org/web/20160413172707/http://blog.wuntee.sexy/CVE-2015-3795/

> But doesn't Apple say that the format isn't guaranteed to stay consistent across releases? Much like how you're not supposed to make system calls yourself; Apple doesn't guarantee compatibility between releases so you're supposed to use the API they provide in system libraries. Programs aren't supposed have their own libxpc, so the format we use for messages shouldn't matter. Our libxpc is supposed to be the one doing both the packing and the unpacking for all programs, right?

This is all true; most well-behaved programs should not rely on XPC wire format, and it does change with OS versions. But all else being equal, it's better to use the very same format Apple uses than not. Tools that dump/inspect XPC communications (if they exist...) are likely going to expect the format to be Apple's, and RemoteXPC may see more usage in the future.

> Or did Apple use a different approach (i.e. statically linking their libxpc into programs) in the PowerPC days?

No; there was no XPC in the PowerPC days at all, they introduced XPC after switching to Intel. But programs that run emulated in Rosetta (so, PowerPC) would still use XPC as it's a part of macOS/Darwin now.

> Then again, it doesn't matter to me what we do here, I just want XPC messaging to work for trustd <-> curl. The `nvlist`/`nvpair` approach was broken, and I thought this was the easiest way to fix it.

Sure, I'm not saying you should use Apple's format right now. Just that we're going to switch to it eventually, so making sure you are aware of that.

Reply:
Ok. 👍